File: Programmeringsuppgift (Java)/Inbetalningar/se/dedev/filetools/Controllers/BetalningsServiceHandler.java
package se.dedev.filetools.Controllers;

import java.io.IOException;
import java.math.BigDecimal;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.text.ParseException;
import java.text.SimpleDateFormat;

import se.dedev.filetools.se.dedev.Pojo.BetalningsServiceData;
import se.dedev.filetools.se.dedev.Pojo.BetalningsServicePayment;
import se.lumera.example.payments.PaymentReceiver;

public class BetalningsServiceHandler implements PaymentReceiver {

    private BetalningsServiceData betalningsServiceData = null;
    private String fileName;
    private List<String> lines;

    // Constructor
    public BetalningsServiceHandler(String Path) {
        setFileName(Path);
        betalningsServiceData = new BetalningsServiceData();
        try {
            retrieveFileLines();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public BetalningsServiceData getBetalningsServiceData() {
        return this.betalningsServiceData;
    }

    private void setFileName(String path) {
        this.fileName = path;
    }

    private String getFileName() {
        return this.fileName;    
    }
    
    private void retrieveFileLines() throws Exception {
        try {
            lines = Files.readAllLines(Paths.get(getFileName()));
        } catch (IOException e) {
            throw e;
        }
        // Öppningspost (Openingpost)
        String l = lines.get(0);
        String kontonummer = l.substring(1, 16);
        String betalningsDatum = l.substring(40, 48);
        SimpleDateFormat sdf = new SimpleDateFormat("yyyymmdd", Locale.ENGLISH);
        Date datum = null;
        try {
            datum = sdf.parse(betalningsDatum);
        } catch (ParseException e) {
            throw new Exception("fel vid datum convertering");
        }
        String valuta = l.substring(48, l.length());
        startPaymentBundle(kontonummer, datum, valuta);

        // Kontroll av Antal poster är korrekt (Checking the correct number of records)
        String poster = l.substring(30, 40).trim();
        if (lines.size() - 1 != Integer.parseInt(poster)) {
            throw new Exception("antal poster stämmer inte (incorrect number of records)");
        }

        // Lops through the remaining lines in the file
        for (int i = 1; i < lines.size(); i++) {
            String line = lines.get(i);
            String bp = line.substring(1, 16).trim();
            try {
                BigDecimal belopp = new BigDecimal(bp.replaceAll(",", "."));
                String ref = line.substring(15, line.length());
                payment(belopp, ref);
            } catch (Exception e) {
                throw new Exception("fel vid skapandet av belopp. Rad:" + i + " (error creating amount. Line: " + i + ")");
            }
        }

        endPaymentBundle();
    }
    
    @Override
    public void startPaymentBundle(String accountNumber, Date paymentDate, String currency) {
        betalningsServiceData.setAccountNumber(accountNumber);
        betalningsServiceData.setCurrency(currency);
        betalningsServiceData.setPaymentDate(paymentDate);
    }

    @Override
    public void payment(BigDecimal amount, String reference) {
        BetalningsServicePayment payment = new BetalningsServicePayment(amount, reference);
        betalningsServiceData.addToPayments(payment);
    }

    @Override
    public void endPaymentBundle() {
        // Perform treatment on the payment data
        new Treatment(betalningsServiceData);
    }
}

File: ProgrammeringsUppgift (NodeJS + React)/ExpressServer/controllers/BetalningsServiceHandler.js
const fs = require('fs');
const path = require('path');
const PaymentReceiver = require('../interface/PaymentReceiver');

/**
 * Parses a file containing payment records in a specific format.
 * @param {string} file_path - The path to the file to parse.
 * @returns {Object} - The parsed payment bundle object.
 */
function BetalningsServiceHandler(file_path) {
  try {
    const fileContent = fs.readFileSync(file_path, 'utf-8');
    const fileLines = fileContent.split('\r\n');

    // Öppningspost (Opening Post)
    const line = fileLines[0];
    const accountNumber = line.substring(1, 16);
    const paymentDateStr = line.substring(40, 48);
    const paymentDate = parseDate(paymentDateStr);
    const formattedPaymentDate = formatDate(paymentDate);
    const currency = line.substring(48, 51);

    PaymentReceiver.startPaymentBundle(
      accountNumber,
      formattedPaymentDate,
      currency
    );

    // Checking the correct number of records
    const poster = line.substring(30, 40).trim();
    if (fileLines.length - 2 !== parseInt(poster)) {
      throw new Error(
        'antal poster stämmer inte (incorrect number of records)'
      );
    }
    // Betalningspost (Payment Post)
    // Loop through the file from index 1 to the second-to-last line
    for (let i = 1; i < fileLines.length - 1; i++) {
      try {
        const amount = parseFloat(
          fileLines[i].substring(1, 15).replace(',', '.')
        );
        const reference = fileLines[i].substring(16, 51).trim();
        PaymentReceiver.payment(amount, reference);
      } catch (error) {
        console.log('Error parsing line:', i);
      }
    }

    // Return the payment bundle object to the Handler
    const paymentBundle = PaymentReceiver.endPaymentBundle();
    return paymentBundle;
  } catch (error) {
    console.error('Error parsing file:', file_path);
    console.error(error);
    throw error;
  }
}

/**
 * Parses a date string and returns a Date object.
 * @param {string} dateStr - The date string to parse.
 * @returns {Date} - The parsed Date object.
 */
function parseDate(dateStr) {
  const year = dateStr.substring(0, 4);
  const month = dateStr.substring(4, 6);
  const day = dateStr.substring(6, 8);
  return new Date(year, month - 1, day);
}

/**
 * Formats a Date object into a specific date format.
 * @param {Date} date - The Date object to format.
 * @returns {string} - The formatted date string.
 */
function formatDate(date) {
  const day = String(date.getDate()).padStart(2, '0');
  const month = String(date.getMonth() + 1).padStart(2, '0');
  const year = date.getFullYear();
  return `${year}${month}${day}`;
}

module.exports = { BetalningsServiceHandler };
